nixos/tests/tandoor-recipes-script-name.nix

```diff
@@ -0,0 +1,40 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
+  name = "tandoor-recipes-script-name";
+
+  nodes.machine = { pkgs, nodes, ... }: {
+    services.tandoor-recipes = {
+      enable = true;
+      extraConfig = {
+        SCRIPT_NAME = "/any/path";
+        STATIC_URL = "${nodes.machine.services.tandoor-recipes.extraConfig.SCRIPT_NAME}/static/";
+      };
+    };
+  };
+
+  testScript = {nodes, ...}:
+  let
+    inherit (nodes.machine.services.tandoor-recipes) address port;
+    inherit (nodes.machine.services.tandoor-recipes.extraConfig) SCRIPT_NAME;
+  in
+  ''
+    username = "any_username"
+    password = "any password"
+
+    cookie_jar_path = "/tmp/cookies.txt"
+    curl = f"curl --cookie {cookie_jar_path} --cookie-jar {cookie_jar_path} --fail --show-error --silent"
+
+    login_url = "http://${address}:${toString port}${SCRIPT_NAME}/accounts/login/"
+
+    # Wait for the service to respond
+    machine.wait_for_unit("tandoor-recipes.service")
+    machine.wait_until_succeeds(f"{curl} {login_url}")
+
+    # Get CSRF token for later requests
+    csrf_token = machine.succeed(f"grep csrftoken {cookie_jar_path} | cut --fields=7")
+
+    machine.succeed(f"DJANGO_SUPERUSER_PASSWORD='{password}' ${pkgs.tandoor-recipes}/lib/tandoor-recipes/manage.py createsuperuser --no-input --username='{username}' --email=nobody@example.com")
```

Comment:
As explained in #287956, you should use `/var/lib/tandoor-recipes/tandoor-recipes-manage` (this might require `sudo` to switch to the `tandoor_recipes` user).

Reply:
That seems to be [undocumented](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22tandoor-recipes-manage%22&type=code). It's also not the name used by the Tandoor manual, so there's no reason any user would realise that.

Reply:
1. It does just wrap `manage.py`.
2. It's the pattern that is used by other Django services in NixOS (e.g: `paperless`). Nextcloud also does a similar thing with its CLI administration tool.
3. The technical reason is to use the same environment variables (i.e: configuration) as the service. 
4. You can improve the documentation if you feel it is unclear.

Reply:
> (this might require `sudo` to switch to the `tandoor_recipes` user).

Does anything actually create the `tandoor_recipes` user? The main Nix file just [sets the service to run as that user](https://github.com/NixOS/nixpkgs/blob/2058a82fb0d09431e0c449bf29ea96c10c821da8/nixos/modules/services/misc/tandoor-recipes.nix#L87), but doesn't create it. And based on the [documentation](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=), it looks like `DynamicUser = true;` means it just creates a random user at runtime.

Reply:
And `machine.succeed("/var/lib/tandoor-recipes/tandoor-recipes-manage migrate")` says "django.db.utils.OperationalError: unable to open database file".

Reply:
> Does anything actually create the tandoor_recipes user?

Systemd does, when the service is running.

I have recently [uploaded a change](#298168) (inspired by Photoprism's module) to attach to the service's namespace to make this work reliably. Something like `sudo /var/lib/tandoor-recipes/tandoor-recipes-manage` should work (you need the `sudo` to be able to read/execute the file IIRC).

Reply:
> And `machine.succeed("/var/lib/tandoor-recipes/tandoor-recipes-manage migrate")` says "django.db.utils.OperationalError: unable to open database file".

So I think your actual issue is not using `sudo /.../tandoor-recipes-manage`. I don't remember having any issues using the manage script on my server. Let me know if you have more issues. 

In trying it again locally, I noticed something.

I think I found an issue in my code, with the following message:

```
/var/lib/tandoor-recipes/tandoor-recipes-manage: line 16: UID: readonly variable
```

So it's not using `tandoor_recipes`' `UID`. I _think_ that's fine as it's probably just going to use `root`'s instead, which should have all the correct rights for DB access etc (as I said, I didn't have any issues when I used it)... But that's still not optimal.

Reply:
I can't find any reference to a `--wdns` argument; it doesn't seem to be available for the manage script at least:

```
usage: manage.py createsuperuser [-h] [--username USERNAME] [--noinput]
                                 [--database DATABASE] [--email EMAIL]
                                 [--version] [-v {0,1,2,3}]
                                 [--settings SETTINGS]
                                 [--pythonpath PYTHONPATH] [--traceback]
                                 [--no-color] [--force-color] [--skip-checks]
```

Reply:
It's a [`nsenter`](https://github.com/NixOS/nixpkgs/blob/e9ee548d90ff586a6471b4ae80ae9cfcbceb3420/nixos/modules/services/misc/tandoor-recipes.nix#L24) argument which is missing. I'm asking you to try adding it locally (to the module) and see if it fixes the SQLite issue.

Reply:
> It's a [`nsenter`](https://github.com/NixOS/nixpkgs/blob/e9ee548d90ff586a6471b4ae80ae9cfcbceb3420/nixos/modules/services/misc/tandoor-recipes.nix#L24) argument which is missing. I'm asking you to try adding it locally (to the module) and see if it fixes the SQLite issue.

Same issue:

```
❯ git diff
diff --git a/nixos/modules/services/misc/tandoor-recipes.nix b/nixos/modules/services/misc/tandoor-recipes.nix
index a2210f3d7db5..f3545fbbf530 100644
--- a/nixos/modules/services/misc/tandoor-recipes.nix
+++ b/nixos/modules/services/misc/tandoor-recipes.nix
@@ -22,7 +22,7 @@ let
     ${lib.toShellVars env}
     eval "$(${config.systemd.package}/bin/systemctl show -pUID,GID,MainPID tandoor-recipes.service)"
     exec ${pkgs.util-linux}/bin/nsenter \
-      -t $MainPID -m -S $UID -G $GID \
+      -t $MainPID -m -S $UID -G $GID --wdns="${pkg}" \
       ${pkg}/bin/tandoor-recipes "$@"
   '';
 in
diff --git a/nixos/tests/tandoor-recipes-script-name.nix b/nixos/tests/tandoor-recipes-script-name.nix
index f48695c38c53..5a14a219f9c2 100644
--- a/nixos/tests/tandoor-recipes-script-name.nix
+++ b/nixos/tests/tandoor-recipes-script-name.nix
@@ -22,6 +22,7 @@ import ./make-test-python.nix (
         inherit (nodes.machine.services.tandoor-recipes.extraConfig) SCRIPT_NAME;
       in
       ''
+        machine.shell_interact()
         username = "any_username"
         password = "any password"
```

```
$ cat /var/lib/tandoor-recipes/tandoor-recipes-manage
#!/nix/store/agkxax48k35wdmkhmmija2i2sxg8i7ny-bash-5.2p26/bin/bash
set -o allexport # Export the following env vars
DEBUG='0'
DEBUG_TOOLBAR='0'
GUNICORN_CMD_ARGS='--bind=localhost:8080'
MEDIA_ROOT='/var/lib/tandoor-recipes'
SCRIPT_NAME='/any/path'
STATIC_URL='/any/path/static/'
eval "$(/nix/store/cym39cl4v79bmbng4bvafkxagz96f4hw-systemd-255.6/bin/systemctl show -pUID,GID,MainPID tandoor-recipes.service)"
exec /nix/store/w8p87ilgm572kw000sbfqvfxziiw2sqz-util-linux-2.39.4-bin/bin/nsenter \
  -t $MainPID -m -S $UID -G $GID --wdns="/nix/store/gdv5fj6b33y8575wd9kfr9yaa0ng43x1-python3.11-tandoor-recipes-1.5.16" \
  /nix/store/gdv5fj6b33y8575wd9kfr9yaa0ng43x1-python3.11-tandoor-recipes-1.5.16/bin/tandoor-recipes "$@"
```

```
$ sudo DJANGO_SUPERUSER_PASSWORD='any password' /var/lib/tandoor-recipes/tandoor-recipes-manage createsuperuser --no-input --username='any_username' --email=nobody@example.com
machine # [  251.110671] sudo[975]:     root : TTY=hvc0 ; PWD=/tmp ; USER=root ; ENV=DJANGO_SUPERUSER_PASSWORD=any password ; COMMAND=/var/lib/tandoor-recipes/tandoor-recipes-manage createsuperuser --no-input --username=any_username --email=nobody@example.com
machine # [  251.156655] sudo[975]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
machine # /var/lib/tandoor-recipes/tandoor-recipes-manage: line 10: UID: readonly variable
machine # Traceback (most recent call last):
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
machine #     self.connect()
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
machine #     return func(*args, **kwargs)
machine #            ^^^^^^^^^^^^^^^^^^^^^
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/db/backends/base/base.py", line 270, in connect
machine #     self.connection = self.get_new_connection(conn_params)
machine #                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
machine #     return func(*args, **kwargs)
machine #            ^^^^^^^^^^^^^^^^^^^^^
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/db/backends/sqlite3/base.py", line 180, in get_new_connection
machine #     conn = Database.connect(**conn_params)
machine #            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
machine # sqlite3.OperationalError: unable to open database file
```

Trying to `cd /var/lib/tandoor-recipes` first produced the same result.